### PR TITLE
RA-1599 : 'Core Apps' DiagnosesFragmentController search to allow dia…

### PR DESCRIPTION
**Pull request by Ruhanga**
Ticket : https://issues.openmrs.org/browse/RA-1599

Missing tests. I will add tests in the next commit.

### CODE REVIEW

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            diagnosis3.setFullySpecifiedName(new ConceptName("Liver cancer", locale));
+            diagnosis3.setConceptClass(conceptService.getConceptClassByName("Diagnosis"));
+            diagnosis3.setDatatype(conceptService.getConceptDatatypeByName("N/A"));
+            diagnosis3.addConceptMapping(new ConceptMap(new ConceptReferenceTerm(source,"5002","Test"), conceptMapType));
+            diagnosis3 = conceptService.saveConcept(diagnosis3);
+            Concept diagnosis4 = new Concept();
+            diagnosis4.setFullySpecifiedName(new ConceptName("Lung cancer infection", locale));
+            diagnosis4.setConceptClass(conceptService.getConceptClassByName("Diagnosis"));
+            diagnosis4.setDatatype(conceptService.getConceptDatatypeByName("N/A"));
+            diagnosis4.addConceptMapping(new ConceptMap(new ConceptReferenceTerm(source,"5003","Test"), conceptMapType));
```
Comment (mks-d) [inline omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: `"J85.2"`

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        }
+
+    }
+
+    @Test
+    public void search_shouldSearchForDiagnosisConceptsFromSpecifiedSets() throws Exception {
+        // setup
+        when(context.getLocale()).thenReturn(locale);
+        String queryString = "infection";
+        String diagnosisSetUuids = LUNG_DISEASE_DIAGNOSIS_SET_UUID;
```
Comment (mks-d) [inline omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: `diagnosisSetUuid` (not plural)

File: omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentController.java [hunk 3]
```diff
@@ -54,11 +54,25 @@
...
                                      @SpringBean("emrConceptService") EmrConceptService emrConceptService,
+                                     @SpringBean("conceptService") ConceptService conceptService,
                                      @RequestParam("term") String query,
+                                     @RequestParam(value = "diagnosisSetUuids", defaultValue = "") String diagnosisSetUuids,
                                      @RequestParam(value = "start", defaultValue = "0") Integer start,
                                      @RequestParam(value = "size", defaultValue = "50") Integer size) throws Exception {
 
-        Collection<Concept> diagnosisSets = emrApiProperties.getDiagnosisSets();
+        Collection<Concept> diagnosisSets = new ArrayList<Concept>();
+        if (StringUtils.isNotEmpty(diagnosisSetUuids) && !"undefined".equalsIgnoreCase(diagnosisSetUuids)) {
```
Comment (mks-d) [inline omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentController.java]: The check on `undefined` shouldn't be here. It's rather a client-side effort to ensure that a valid string is sent here. If `undefined` lands here anyway then so be it.

Comment (Ruhanga) [reply omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentController.java]: Yes, this is actually [catered for on RA-1603](https://github.com/openmrs/openmrs-module-coreapps/pull/199/files#diff-7d6c8132f77707ede46ad8ec6a618a59R78) but some js implementations may consider an **empty string** as `undefined`.

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java (hunk 2 above)
Comment (Ruhanga) [reply omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: I've addressed this.

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        administrationService.saveGlobalProperty(new GlobalProperty(EmrApiConstants.EMR_CONCEPT_SOURCES_FOR_DIAGNOSIS_SEARCH,
+                "ICD-10-WHO"));
+
+        // Set up 'Lung diseases diagnoses' set
+        {
+            Concept diagnosis1 = new Concept();
+            diagnosis1.setFullySpecifiedName(new ConceptName("Asthma", locale));
+            diagnosis1.setConceptClass(conceptService.getConceptClassByName("Diagnosis"));
+            diagnosis1.setDatatype(conceptService.getConceptDatatypeByName("N/A"));
+            diagnosis1.addConceptMapping(new ConceptMap(new ConceptReferenceTerm(source,"5000","Test"), conceptMapType));
```
[the remark this answers is not in the document]
Comment (Ruhanga) [reply omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: Thanks @mks-d, I've addressed this and related concerns below.

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+public class DiagnosesFragmentControllerTest extends BaseModuleWebContextSensitiveTest { 
+
+    private String DIAGNOSIS_SET_OF_SETS_UUID = "63b378a5-5065-11de-80cb-0018ea23ee37";
+
+    private String CONCEPT_SOURCE_UUID = "75f5b378-5065-11de-80cb-001e378eb67e";
+    
+    private String LUNG_DISEASE_DIAGNOSIS_SET_UUID = "d1010973-803e-8659-4415-c01707c01dec";
+
+    private String CANCEROUR_DISEASE_DIAGNOSIS_SET_UUID = "d710b7h4-40b7-d333-b449-6e0e15d0739d";
```
[the remark this answers is not in the document]
Comment (Ruhanga) [reply omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: Nice catch,  I have corrected the typo. :)

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+    private UiSessionContext context;
```
[the remark this answers is not in the document]
Comment (Ruhanga) [reply omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: I've resolved this.

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+            diagnosisSuperSet.addSetMember(conceptService.getConceptByUuid(CANCEROUR_DISEASE_DIAGNOSIS_SET_UUID));
+            conceptService.saveConcept(diagnosisSuperSet);
+        }
+
+    }
+
+    @Test
+    public void search_shouldSearchForDiagnosisConceptsFromSpecifiedSets() throws Exception {
+        // setup
+        when(context.getLocale()).thenReturn(locale);
```
[the remark this answers is not in the document]
Comment (Ruhanga) [reply omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: Yes, this should be in the setup method. Resolved.

File: omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentController.java (hunk 3 above)
Comment (mks-d) [reply omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentController.java]: So please remove that check, it not the place of a Java controller to take care of such JavaScript edge cases.

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+        // verify
+        assertEquals(2, conceptSearchResults.size());
+        assertThat((String) ((SimpleObject) conceptSearchResults.get(0).get("conceptName")).get("name"), is("Lung cancer infection"));
+        assertThat((String) ((SimpleObject) conceptSearchResults.get(1).get("conceptName")).get("name"), is("Tuberculosis infection"));
+
+    }
+
+    @Test
+    public void search_shouldSearchForDiagnosisConceptsUsingGloballyDefinedSuperSetGivenUndefinedDiagnosisSetUuids() throws Exception {
```
Comment (mks-d) [inline omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: So this test should eventually go.

File: omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentController.java (hunk 3 above)
Comment (Ruhanga) [reply omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentController.java]: Yes, I have cleared this.

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java (hunk 8 above)
Comment (Ruhanga) [reply omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: I've resolved this.

File: omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+import org.springframework.beans.factory.annotation.Qualifier;
+
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class DiagnosesFragmentControllerTest extends BaseModuleWebContextSensitiveTest {
```
Comment (mks-d) [inline omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: We just need a good old JUnit test that shows that we go either way depending on whether the new attribute is present or not, well formatted or not... etc.

Something for example that shows, amongst other things, that `emrApiProperties.getDiagnosisSets()` is _never_ called when the overriding attribute is provided.

Do you agree?

Comment (Ruhanga) [reply omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/DiagnosesFragmentControllerTest.java]: Okay, yes really. This actually makes the test cases a hole lot easier

**Comment by Ruhanga** [comment]
@mks-d, @mogoodrich, is this fine to merge?
